contracts/helpers/midas/MidasRedemptionVaultGateway.sol

```diff
@@ -0,0 +1,226 @@
+
+        pendingRedemptions[msg.sender] = PendingRedemption({
+            isActive: true,
+            isManuallyCleared: false,
+            requestId: requestId,
+            timestamp: block.timestamp,
+            remainder: 0
+        });
+    }
+
+    /// @notice Withdraws tokens from a fulfilled redemption request
+    /// @param amount Amount of output token to withdraw
+    /// @dev Supports partial withdrawals by tracking remainder
+    function withdraw(uint256 amount) external nonReentrant {
+        if (amount == 0) {
+            revert ZeroAmountException();
+        }
+
+        PendingRedemption memory pending = pendingRedemptions[msg.sender];
+
+        if (!pending.isActive) {
+            revert NoPendingRedemptionException();
+        }
+
+        (
+            address sender,
+            address tokenOut,
+            uint8 status,
+            uint256 amountMTokenIn,
+            uint256 mTokenRate,
+            uint256 tokenOutRate
+        ) = IMidasRedemptionVault(midasRedemptionVault).redeemRequests(pending.requestId);
+
+        if (sender != address(this)) {
+            revert InvalidRequestException();
+        }
+
+        if (status != 1 && !pending.isManuallyCleared) {
+            revert RedemptionNotFulfilledException();
+        }
+
+        uint256 availableAmount;
+
+        if (pending.remainder > 0) {
+            availableAmount = pending.remainder;
+        } else {
+            availableAmount = _calculateTokenOutAmount(amountMTokenIn, mTokenRate, tokenOutRate, tokenOut);
+        }
+
+        if (amount > availableAmount) {
+            revert AmountExceedsAvailableException();
+        }
+
+        if (amount == availableAmount) {
+            delete pendingRedemptions[msg.sender];
+        } else {
+            pendingRedemptions[msg.sender].remainder = availableAmount - amount;
+        }
+
+        IERC20(tokenOut).safeTransfer(msg.sender, amount);
+    }
+
+    /// @notice Returns the expected amount of output token for a account's pending redemption
+    /// @param account account address to check
+    /// @param tokenOut Output token to check
+    /// @return Expected amount of output token, considering any partial withdrawals
+    function pendingTokenOutAmount(address account, address tokenOut) external view returns (uint256) {
+        PendingRedemption memory pending = pendingRedemptions[account];
+
+        if (!pending.isActive) {
+            return 0;
+        }
+
+        (address sender, address requestTokenOut,, uint256 amountMTokenIn, uint256 mTokenRate, uint256 tokenOutRate) =
+            IMidasRedemptionVault(midasRedemptionVault).redeemRequests(pending.requestId);
+
+        if (sender != address(this) || requestTokenOut != tokenOut) {
+            return 0;
+        }
+
+        if (pending.remainder > 0) {
+            return pending.remainder;
+        } else {
+            return _calculateTokenOutAmount(amountMTokenIn, mTokenRate, tokenOutRate, tokenOut);
```

Comment:
As discussed in private, I kind of hate the idea that once a user initiates a withdrawal, we immediately exclude all risks related to Midas/underlying protocol when pricing their position, effectively freezing the `mTokenRate`.

For example, while not fully resolving the issue (as there exist `MidasRedemptionVault.safeBulkApproveRequest` variants that use either saved or admin-specified rate), using the current `mTokenRate` from the `mTokenDataFeed` allows to capture at least some exposure to mToken.